mixer/src/migrate/index.ts

```
@@ -0,0 +1,12 @@
+import client from "../db";
+
+// Currently we only have a few migration so I don't bring in extra dependencies to manage it
+// when we add a dozen more, switch to db-mate
+export const migrate = async () => {
+  await Promise.all([
+    client.query('alter table turing.blocks add column if not exists collator_id varchar(255)'),
+    client.query('alter table turing.blocks add column if not exists state_root varchar(255)'),
+    client.query('alter table turing.blocks add column if not exists extrinsics_root varchar(255)'),
+    client.query('create index concurrently if not exists blocks_collator on turing.blocks (collator_id);')
+  ]);
+}
```

I’m curious, what happens if an error happens within migrate code. Would it stop the entire problem. In other words, does the program have any catch all exceptions code at root level?

This function doesn’t return anything. I am not sure if you need it, but `return Promise.all()`; could return the result of the 4 promises to the caller.

---

> I’m curious, what happens if an error happens within migrate code. Would it stop the entire problem. In other words, does the program have any catch all exceptions code at root level?

It will just throw unhandle promise and coninue on unfortunately. I overlook it. Let me add an unhandle error at the node run time to catch it and crash out. better to be error and stop than in an undefined state.

I overlook cuz i's unlikely to cause error because the SQL is idempotent, and if it's db error, it will crash out at later phase.

when we're handle more and more, I think we can connect with a properly migration system such as https://github.com/amacneil/dbmate. We need it because SubQL doesn't allow to change an existing schema.

---

> This function doesn’t return anything. I am not sure if you need it, but return Promise.all(); could return the result of the 4 promises to the caller.

Yes, we don't need the. return. if they return they are succeed. If they error out we got an error back and crash early like this

```
/Users/vinh/Sites/oak/oak-subql/mixer/dist/index.js:93
    process.on('unhandledRejection', error => { throw error; });
                                                ^

error: relation "turing.blocksaa" does not exist
    at Parser.parseErrorMessage (/Users/vinh/Sites/oak/oak-subql/mixer/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/Users/vinh/Sites/oak/oak-subql/mixer/node_modules/pg-protocol/dist/parser.js:126:29)
```

I group ones that can run in-dependently from each other into `Promise.all` so we don't need to wait for them sequentially.

---

okay, if you are adding a catch all at the root level, we might want a sentry hook there as well, similar to https://github.com/OAK-Foundation/webapp-parse-backend/blob/d6c0e97b3039654b2a6ffffb27da0341f111e000/src/index.js#L96

That way, Sentry can capture and keep all the issues and we can review them when we need to.

---

Yea, I added Sentry monitoring. Example https://oak-network.sentry.io/issues/4223227897/?project=4505288232665088&query=is%3Aunresolved&referrer=issue-stream&stream_index=0 (fake an error with invalid table) (Still need to config source map)